### ISSUE TITLE
fileName for Types doesnt work

### ISSUE THREAD

**Issue by kakuschke** [dmitryd/typo3-realurl#60]
I have some XML and other Pages in the fileName Section.

Example:
```
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['realurl']['_DEFAULT'] = array(
	'fileName' => array (
		'index' => array(
			'robots.txt' => array(
				'keyValues' => array(
					'type' => 841231,
				),
			),
			'sitemap.xml' => array(
				'keyValues' => array(
					'type' => 841132,
					'page' => 1,
					'id' => 1,
					'no_cache' => 1,
				),
			),
		),
	),
);
```
Without RealUrl and with RealUrl 1.x it works. Even the Example from this Documentation does not work.

Please HELP!

**Comment by dmitryd** [comment]
> fileName for Types doesnt work

I really cannot fix "does not work" because this does not tell me anything. What exactly happens? Does the computer die? Do you see any exceptions? Do you see errors?

> Without RealUrl and with RealUrl 1.x it works

Do you mean the file name work without realurl?

**Comment by kakuschke** [comment]
The URL example.com/robots.txt or example.com/sitemap.xml have a Error 404. That is mean with "doesnt work".

**Comment by dmitryd** [comment]
Is it a TYPO3 404 error?

**Comment by kakuschke** [comment]
Yes a 404 Error.

**Comment by kakuschke** [comment]
example.com/index.php?id=1&type=841231 show the robots.txt File korrekt

**Comment by kakuschke** [comment]
example.com/robots.txt with RealUrl 1.x shows the robots.txt File korrekt as well

**Comment by kakuschke** [comment]
But my Computer does not die! ;-)

**Comment by dmitryd** [comment]
Is it a TYPO3 404 error or your web server 404 error?

**Comment by kakuschke** [comment]
It displays the configured page for 404Errors (in Install Tool => "[FE][pageNotFound_handling] = index.php?id=34" ).
What is the different, I thought that TYPO3 just handle the Webserver Error? Sorry

**Comment by dmitryd** [comment]
Could you temporarily remove that setting for `pageNotFound_handling` and see what TYPO3 tells you? It will give you a specific error message.

**Comment by kakuschke** [comment]
"[FE][pageNotFound_handling] = " value is now empty. The result is a 404 Error as well, but with the TYPO3 404Error Page.

The message is:
"Page Not Found
Reason: Segment "" was not a keyword for a postVarSet as expected on page with id=1."

**Comment by dmitryd** [comment]
Currently unable to reproduce :(

**Comment by kakuschke** [comment]
Many thanks sofar, but I think it will be fixed ;-)

**Comment by dmitryd** [comment]
I keep it open in case of anybody else have the issue too. Than we can figure out what is wrong. I tried various ways to reproduce it but could not. Possibly it is related to configuration or page tree structure or page settings. I am not sure.

Btw, for the sitemap, you will not be able to set vars as they are described in your configuration because `id` will always be set by realurl on the later stage. I think you do not really need it there :smile:

**Comment by kakuschke** [comment]
Has something changed in the .htaccess?

**Comment by dmitryd** [comment]
No.

**Comment by Gorli** [comment]
I can confirm this problem!
 
For example: http: //localhost/robots.txt
Here is an error message
"Reason: Segment" "was not a keyword for a postVarSet as expected on page with id = 1".

When I go to another level - for example, http: //localhost/komponenten/robots.txt

then it works ?! The robots.txt is displayed! Until realurl version 1.13.6 there was NEVER problems !!!

I can be the cause not locate and think there is an error in the new version of realurl ?!

thank you

Harald

**Comment by dmitryd** [comment]
Guys, please, post your full realurl config.

**Comment by Gorli** [comment]
Hello!

I send my complete realurl_conf. However, I had the file only after realurl_conf.php realurl_conf.txt rename so I can upload it.

[realurl_conf.txt](https://github.com/dmitryd/typo3-realurl/files/163072/realurl_conf.txt)
The typoscript I send with as file.
[robots.txt](https://github.com/dmitryd/typo3-realurl/files/163089/robots.txt)


Thank you very much

Harald

**Comment by dmitryd** [comment]
@Gorli Does any other file name (like `atom.xml`) work?

**Comment by Gorli** [comment]
I can not say there is not currently in use unfortunately. In use are currently sitemap.xml (typo3 105) and robots.txt (type 106). In both there is a problem.

Page Not Found
Reason: Segment "" was not a keyword for a postVarSet as expected on page with id =. 1

I am also feverishly on Search. Unfortunately no solution.

**Comment by Gorli** [comment]
Stupid question.
Is the point after id in order?

with id=1.

**Comment by Gorli** [comment]
I send a picture
![error](https://cloud.githubusercontent.com/assets/4998028/13601451/8692a60c-e530-11e5-9290-3fae5c3dfd32.PNG)

**Comment by dmitryd** [comment]
Are you using latest realurl (2.0.5)?

**Comment by Gorli** [comment]
Yes

**Comment by kakuschke** [comment]
Hi Dimitry,
is your script from the current docu correct working? 

Script means:
https://github.com/dmitryd/typo3-realurl/wiki/Configuration-reference#filenameindexkeyvalues

'fileName' => array (
    'index' => array(
        'print.html' => array(
            'keyValues' => array(
                'type' => 98
            )
        ),
        'rss.xml' => array(
            'keyValues' => array(
                'type' => 76
            )
        )
    )
)

Cheers
Thomas

**Comment by kakuschke** [comment]
I can confirm that the problem appears NOT on the next level.

example.com/robots.txt - dont work
example.com/level1/robots.txt - is working

doesnt matter what filename it is.

Adding my real_conf.php (as txt
[realurl_conf.txt](https://github.com/dmitryd/typo3-realurl/files/163333/realurl_conf.txt)
)

**Comment by dmitryd** [comment]
I finally was able to reproduce and fix it.

**Comment by kakuschke** [comment]
yipie, Thanks

**Comment by kakuschke** [comment]
Ok, the fileName is working correct now.

**Comment by kakuschke** [comment]
But in the rootpage are 2 Slashes.

example.com// (1 too mutch)

**Comment by kakuschke** [comment]
You want a new issue for the "doubleslash thing"

**Comment by christophe-nc** [comment]
I had a similar issue with TYPO3 6.2.21 since updating to realURL 2.x
I've solved the problem in modifying function decodePostVarSets. I just added a verification to verify if $postVarSetKey is not empty : 
`if ($postVarSetKey!="") {
					if (!isset($postVarSets[$postVarSetKey]) || !is_array($postVarSets[$postVarSetKey])) {
						$this->handleNonExistingPostVarSet($pageId, $postVarSetKey, $pathSegments);
					} else {
						$postVarSetConfiguration = $postVarSets[$postVarSetKey];
						// Note: we do not support aliases for postVarSets!
						if (is_array($postVarSetConfiguration)) {
							foreach ($postVarSetConfiguration as $postVarConfiguration) {
								$this->decodeSingleVariable($postVarConfiguration, $pathSegments, $requestVariables, $previousValue);
							}
						}
					}
				}`

**Comment by axsc** [comment]
I have a similar problem with Typo3 7.2.9 and realURL 2.0.14.

The url /?type=841132 works.
The url /sitemap.xml only works if i am logged in in the backend otherwise i get "page does not exist".
My realURL configuration is similar to kakuschke.

Ist this  problem already fixed in the new version 2.1.0 ?
